ark/phenotyping/cluster_helpers.py

```diff
@@ -1,16 +1,310 @@
+from abc import ABC, abstractmethod
 import bisect
 import feather
 from itertools import combinations
 import numpy as np
+import os
 import pathlib
 import pandas as pd
+from pyFlowSOM import map_data_to_nodes, som
 from scipy.stats import zscore
 from sklearn.cluster import AgglomerativeClustering
 from typing import Callable, List, Protocol, runtime_checkable
+import warnings
 
+from ark.utils.io_utils import list_files, validate_paths
 from ark.utils.misc_utils import verify_in_list
 
 
+class PixieSOMCluster(ABC):
+    @abstractmethod
+    def __init__(self, weights_path: pathlib.Path, columns: List[str], num_passes: int = 1,
+                 xdim: int = 10, ydim: int = 10, lr_start: float = 0.05, lr_end: float = 0.01):
+        """Generic implementation of a pyFlowSOM runner
+
+        Args:
+            weights_path (pathlib.Path):
+                The path to save the weights to.
+            columns (List[str]):
+                The list of columns to subset the data on.
+            num_passes (int):
+                The number of SOM training passes to use.
+            xdim (int):
+                The number of SOM nodes on the x-axis.
+            ydim (int):
+                The number of SOM nodes on the y-axis.
+            lr_start (float):
+                The initial learning rate.
+            lr_end (float):
+                The learning rate to decay to
+        """
+        self.weights_path = weights_path
+        self.weights = None if not os.path.exists(weights_path) else feather.read_dataframe(
+            weights_path
+        )
+        self.columns = columns
+        self.num_passes = num_passes
+        self.xdim = xdim
+        self.ydim = ydim
+        self.lr_start = lr_start
+        self.lr_end = lr_end
+
+    @abstractmethod
+    def normalize_data(self) -> pd.DataFrame:
+        """Generic implementation of the normalization process to use on the input data
+
+        Returns:
+            pandas.DataFrame:
+                The data with `columns` normalized by the values in `norm_data`
+        """
+        pass
+
+    def train_som(self, data: pd.DataFrame):
+        """Trains the SOM on the data provided and saves the weights generated
+
+        Args:
+            data (pandas.DataFrame):
+                The input data to train the SOM on.
+        """
+        # make sure to run a deterministic SOM for reproducibility purposes
+        som_weights = som(
+            data=data.values, xdim=self.xdim, ydim=self.ydim,
+            alpha_range=(self.lr_start, self.lr_end),
+            deterministic=True
```

Comment:
Hm, so this `deterministic=True` thing is bothering me slightly. In Zak's implementation, he has these lines: https://github.com/angelolab/pyFlowSOM/blob/c7989bdaaaa2f3d23c936df4c0b64c2769198353/pyFlowSOM/cyFlowSOM.pyx#L106-L107

So if `deterministic=True`, that means the nodes are not randomly initialized (it's always the first 100 data points). Perhaps with a lot of iterations, it doesn't matter how the nodes are initialized, but we never tested this. In the og implementation of FlowSOM, it's always random (unless the user specifies the initial nodes): https://github.com/SofieVG/FlowSOM/blob/d5fc07fd81ff1840a4f8791e0e7526d86864cf1a/R/2_buildSOM.R#L190.

I think it would be better to always have a true random initialization of nodes. I think Zak probably made this `deterministic=True` feature for testing. I think in practice, it would still be better to set a seed, and have that be something the users can change if they wish. This would require changing the pyFlowSOM code. I don't think we need this `deterministic` parameter - instead, we should replace it with a `seed` parameter, and just set the seed before initializing the nodes (before this line: https://github.com/angelolab/pyFlowSOM/blob/c7989bdaaaa2f3d23c936df4c0b64c2769198353/pyFlowSOM/cyFlowSOM.pyx#L109). 

Here, if `deterministic=True`, Zak sets a random seed: https://github.com/angelolab/pyFlowSOM/blob/c7989bdaaaa2f3d23c936df4c0b64c2769198353/pyFlowSOM/cyFlowSOM.pyx#L146. We could probably just replace that with whatever seed the user chooses.

I don't think this is that complex (I think @alex-l-kong or @srivarra could do it?). @ngreenwald thoughts?

Reply:
Or alternatively, set the seed before the call to pyFlowSOM and then set `deterministic=False` in the function call. We'd have to confirm that this actually does make pyFlowSOM reproducible though.

Reply:
@cliu72 There are a couple of ways we can do this:

1. In [Python / Cython](https://numpy.org/doc/stable/reference/random/generator.html#numpy.random.default_rng)
2. Directly in C with the [Numpy C - API](https://numpy.org/devdocs/reference/random/c-api.html)

Reply:
@cliu72 let me test this out. However, if the actual subsetting process is set to a random seed (which it is in `preprocess_fov` prior to calling `create_fov_pixel_data`), we always should be getting the same first 100 rows right?

Reply:
@cliu72 I just tested it out. Setting `np.random.seed` in Python prior to the call to `pyFlowSOM.som` does not randomly initialize the weights. I think this random seed needs to be explicitly set in C.

Using `deterministic=True` does ensure the weights are randomly initialized because the random seed setting is passed off to C. Also it looks like these nodes are initialized to random N data points of `data`:

```
    if nodes is None:
        # If we don't supply nodes, we randomly sample them from the data
        if deterministic == True:
            nodes = data[0: xdim * ydim, :]
        else:
            nodes = data[np.random.choice(data.shape[0], xdim * ydim, replace=False), :]  # this will randomly select `xdim * ydim points from data
```

Reply:
I think it'll be worthwhile for Sri and/or Alex to see what it takes to make an updated release for pyFlowSOM, to make sure we can handle it when things come up in the future. Changing the logic would be quite straightforward. I'm not that familiar with C, so I can't quite figure out where this random seed gets translated to the C_seed, and how that actually impacts the nodes.

Reply:
Okay so I just did this test:
```
np.random.seed(100)
pyFlowSOM.som(dat.values, deterministic=False)

np.random.seed(100)
pyFlowSOM.som(dat.values, deterministic=False)
```

And those were returning a different output - meaning that it does indeed randomly initialize the weighs, but setting a random seed isn't doing anything (not what we want). What we want is for those two functions to return the same output.

For this, I think the easiest thing to try is to change cyFlowSOM.pyx itself and not touch the c code (although I will open another issue in the pyFlowSOM repository in a sec that would need to change c code...unrelated to this though). The og version of FlowSOM did not set a random seed in the c code (the random seed was set in R before calling the FlowSOM function). Here, I think we can try adding `np.random.seed(seed)` to the beginning of the som function in cyFlowSOM.pyx (and add the seed as a parameter to the som function). If after testing that with `deterministic=False` and confirming that setting the same seed returns the same output and changing the seed returns a different output, I think we can completely remove having the `deterministic` parameter.

@alex-l-kong Regarding your point that the subsetting process is using a random seed - yes that is true, but I think we should deal with the randomization here independent of the input data fed into the SOM. There could be the case with small datasets (like mine), where I am not subsetting at all. In that case, taking the first N rows of the data might be like the first row of pixels in the image (which are not very diverse). I think we want to keep the functionality of being able to feed the SOM the same exact input data with a different seed and get different results. Overall, I think it would be better to try to recapitulate as much of the original FlowSOM code as possible - and that includes randomizing the nodes (independent of what the input data is).

Reply:
Ugh actually, looking at this a bit more, Zak changed the c code from the original package. In the original c code, https://github.com/SofieVG/FlowSOM/blob/master/src/som.c, there are these lines:

https://github.com/SofieVG/FlowSOM/blob/d5fc07fd81ff1840a4f8791e0e7526d86864cf1a/src/som.c#L14-L15
https://github.com/SofieVG/FlowSOM/blob/d5fc07fd81ff1840a4f8791e0e7526d86864cf1a/src/som.c#L93

It seems like these were removed by Zak. I think we would need to add those lines back in, and adding these back in would allow us to use np.random.seed.

Reply:
From my initial analysis I think the only reason `GetRNGstate` and `PutRNGstate` are called in SofieVG's implementation is because they make an external call to `R_CheckUserInterrupt` every 1024 updates, as well as R's internal `unif_rand` function for random selections. These functions are needed to preserve the original default seed since C is actually calling an R-based package (see https://www.rdocumentation.org/packages/httpuv/versions/1.6.0/topics/getRNGState). However, Zak implemented this logic using just C (https://github.com/angelolab/pyFlowSOM/blob/c7989bdaaaa2f3d23c936df4c0b64c2769198353/pyFlowSOM/flowsom.c#L6-L8), which shouldn't cause any such issues to my knowledge.

It also seems like the `deterministic` param is a recreation of `init` in the `SOM` function of `FlowSOM`, however the deterministic creation of the weights in `FlowSOM` is a lot more sophisticated (https://github.com/SofieVG/FlowSOM/blob/d5fc07fd81ff1840a4f8791e0e7526d86864cf1a/R/2_buildSOM.R#L279-L296), meaning we avoid the issue you brought up of sampling non-diverse pixels. I'm not sure why Zak didn't choose to implement this version and defaulted to a basic top-N sampling for `deterministic=True`.

In either case, I think the main requirement here is ensuring that `pyFlowSOM.som` can take an explicit random seed to initialize the weights in all cases. I'd argue for dropping the `deterministic` parameter and instead pass a `seed` parameter that is then set towards the beginning. In this way, we should be able to use something like:

```
if nodes is None:
    np.random.seed(seed)  # Sri does make a good point about using np.random.default_rng instead per https://albertcthomas.github.io/good-practices-random-number-generators/, this can be a PR on its own
    nodes = data[np.random.choice(data.shape[0], xdim * ydim, replace=False), :]
```

to generate the same starting nodes each time on the same dataset.

Reply:
I told Zak that we never use the `init` parameter and to leave it out of his implementation. In the original version of FlowSOM, the default is `init=FALSE` and we never change that, so the complex initialization function you linked isn't relevant for us (we never use that function).

What we do use is a random initialization of nodes (https://github.com/SofieVG/FlowSOM/blob/d5fc07fd81ff1840a4f8791e0e7526d86864cf1a/R/2_buildSOM.R#L190-191), which is what is currently implemented in Zak's version when `deterministic=False`. I think Zak only implemented `deterministic=True` for testing.

And yes as you outlined and as I talk about above, I think setting a random seed before nodes is the way to go and removing `deterministic`. Your code outline looks good to me.

Reply:
Okay great. Is there anything else besides this issue that still needs to be addressed from this PR from you @cliu72 ? We discussed today that it will be easier to merge this one in using the existing release of pyFlowSOM, and then have a separate PR that fixes the randomization issue in a different pyFlowSOM release.

Reply:
Added a few more small comments